### PR TITLE
Fixes #385: Added swap plugins to snapctl

### LINKED ISSUE #385: Expose SwapPlugins via the CLI

### PR DESCRIPTION
**Pull request by tiffanyfj**
Fixes #385 and #834 

Summary of changes:
- Added swap plugins to snapctl via the client

Testing done:
- Swapping plugins with different types/names
- Swapping plugins with the same type and name
- Swapping a plugin with itself

Still need to add a test for rollback failure and remove swap plugins from control/control_test.go.

@intelsdi-x/snap-maintainers

### CODE REVIEW

File: cmd/snapctl/commands.go [hunk 1]
```diff
@@ -110,6 +110,17 @@ var (
 					},
 				},
 				{
+					Name:   "swap",
+					Usage:  "swap <load_plugin_path> <unload_plugin_type>:<unload_plugin_name>:<unload_plugin_version> or unload -t <unload_plugin_type> -n <unload_plugin_version> -v <unload_plugin_version>",
```
Comment (IRCody) [inline cmd/snapctl/commands.go]: -n should have <unload_plugin_name>

File: cmd/snapctl/commands.go [hunk 2]
```diff
@@ -101,7 +101,7 @@ var (
 				},
 				{
 					Name:   "unload",
-					Usage:  "unload <plugin_type>:<plugin_name>:<plugin_version> or unload -t <plugin_type> -n <plugin_version> -v <plugin_version>",
+					Usage:  "unload <plugin_type>:<plugin_name>:<plugin_version> or unload -t <plugin_type> -n <plugin_name> -v <plugin_version>",
```
Comment (geauxvirtual) [inline cmd/snapctl/commands.go]: This should be a separate commit for this fix tagged to an issue.

File: mgmt/rest/client/plugin.go [hunk 3]
```diff
@@ -78,6 +79,63 @@ func (c *Client) UnloadPlugin(pluginType, name string, version int) *UnloadPlugi
...
+func (c *Client) SwapPlugin(loadPath []string, unloadType, unloadName string, unloadVersion int) *SwapPluginsResult {
+	r := &SwapPluginsResult{}
+
+	if len(loadPath) != 1 {
+		r.Err = errors.New("Did not receive only one plugin to load.")
+		return r
+	}
+	// Check if plugin you are trying to unload is loaded
+	var isLoaded bool
+	lps := c.GetPlugins(false)
```
Comment (geauxvirtual) [inline mgmt/rest/client/plugin.go]: Client needs to be expanded to use GetPlugin from the server. This would return a single plugin or no plugin, and no need to iterate through all plugins in the system.

Comment (tiffanyfj) [inline mgmt/rest/client/plugin.go]: Hey could you elaborate a little? I found
`func (s *Server) getPlugin(w http.ResponseWriter, r *http.Request, p httprouter.Params) ` but not sure if that is what you meant… Thanks

Comment (tiffanyfj) [inline mgmt/rest/client/plugin.go]: Updated.

File: cmd/snapctl/plugin.go [hunk 4]
```diff
@@ -116,6 +116,84 @@ func unloadPlugin(ctx *cli.Context) {
...
+		pDetails = filepath.SplitList(ctx.Args()[1])
+		if len(pDetails) == 3 {
+			pType = pDetails[0]
+			pName = pDetails[1]
+			pVer, err = strconv.Atoi(pDetails[2])
+			if err != nil {
+				fmt.Println("Can't convert version string to integer")
+				cli.ShowCommandHelp(ctx, ctx.Command.Name)
+				os.Exit(1)
+			}
```
Comment (IRCody) [inline cmd/snapctl/plugin.go]: Would it be worthwhile to give an error message here so that inputs like 'collector:mock' don't get the error "Must provide plugin type"?

File: mgmt/rest/client/plugin.go [hunk 5]
```diff
@@ -78,6 +80,54 @@ func (c *Client) UnloadPlugin(pluginType, name string, version int) *UnloadPlugi
...
+			r.Err = errors.New("Failed to rollback after error unloading plugin.")
+		}
+		return r
+	}
+	upr := up.PluginUnloaded
+	r.LoadedPlugin = lp.LoadedPlugins[0]
+	r.UnloadedPlugin = upr
+	return r
+}
+
```
Comment (candysmurf) [inline mgmt/rest/client/plugin.go]: What will happen while unloading a plugin which has running tasks?

Comment (IRCody) [inline mgmt/rest/client/plugin.go]: Dealing with that would happen on the server side I think.

File: cmd/snapctl/plugin.go (hunk 4 above)
Comment (tiffanyfj) [inline cmd/snapctl/plugin.go]: Changed it.

File: cmd/snapctl/commands.go (hunk 2 above)
Comment (tiffanyfj) [inline cmd/snapctl/commands.go]: Missed the tagged to an issue part. Updated.

**Comment by IRCody** [comment]
LGTM

**Comment by jcooklin** [comment]
Let's keep this PR open and expose the swap capability in control through the REST endpoint before merging.  

cc: @geauxvirtual, @tjmcs

File: cmd/snapctl/plugin.go [hunk 6]
```diff
@@ -116,6 +116,88 @@ func unloadPlugin(ctx *cli.Context) {
 	fmt.Printf("Type: %s\n", r.Type)
 }
 
+func swapPlugins(ctx *cli.Context) {
+	// plugin to load
+	pAsc := ctx.String("plugin-asc")
+	var paths []string
+	if len(ctx.Args()) < 1 || len(ctx.Args()) > 2 {
```
Comment (danielscottt) [inline cmd/snapctl/plugin.go]: @tiffanyfj does `Args` return just positional arguments, or all items passed at the command line?  If so, wouldn't a swap with the loaded plugin being loaded with a signing file have a `len` of 4?
```
                      [1]        [2]          [3]            [4]
$ snapctl plugin swap ~/mock2 --plugin-asc ~/mock2.asc collector:mock:1
```

File: cmd/snapctl/plugin.go [hunk 7]
```diff
@@ -116,6 +116,88 @@ func unloadPlugin(ctx *cli.Context) {
...
+		if !strings.Contains(pAsc, ".asc") {
+			fmt.Println("Must be a .asc file for the -a flag")
+			cli.ShowCommandHelp(ctx, ctx.Command.Name)
+			os.Exit(1)
+		}
+		paths = append(paths, pAsc)
+	}
+
+	// plugin to unload
+	var pDetails []string
```
Comment (danielscottt) [inline cmd/snapctl/plugin.go]: nit:
```go
var (
        pDetails     []string
        pType, pName string
        pVer         int
        err          error
)
```

File: cmd/snapctl/plugin.go (hunk 6 above)
Comment (tiffanyfj) [inline cmd/snapctl/plugin.go]: Just positional.
```
./build/bin/snapctl plugin swap build/plugin/snap-collector-mock1 collector:mock:2 -a build/plugin/snap-collector-mock1.asc
len(ctx.Args()) = 2
Plugin loaded
Name: mock
Version: 1
Type: collector
Signed: true
Loaded Time: Sun, 22 May 2016 17:40:11 PDT


Plugin unloaded
Name: mock
Version: 2
Type: collector
```

File: cmd/snapctl/plugin.go (hunk 7 above)
Comment (tiffanyfj) [inline cmd/snapctl/plugin.go]: Seems more legible separate. But I can switch it--would be switching the one for unload too if I did that. I copied it from unload.

**Comment by danielscottt** [comment]
LGTM pending tests!

**Comment by danielscottt** [comment]
ping @geauxvirtual for +2

**Comment by geauxvirtual** [comment]
LGTM
